bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -4595,8 +4595,18 @@ ReplicaImp::ReplicaImp(bool firstTime,
 ReplicaImp::~ReplicaImp() {
   // TODO(GG): rewrite this method !!!!!!!! (notice that the order may be important here ).
   // TODO(GG): don't delete objects that are passed as params (TBD)
-  internalThreadPool.stop();
-  postExecThread_.stop();
+  if (!internalThreadPool.isStopped()) {
+    LOG_WARN(
+        GL,
+        "Internal Thread pool should have already been stopped. Stopping it now but race condition may still occur");
```

Comment:
Before my changes, this is where the thread pool was stopped - which caused the race. now the thread pool is stopped in the "stop" func instead of the destructor. I still left it in the original code in the destructor but added this message to indicate to developers that it shouldn't be done there

Reply:
placed assert instead of warning